File: Data Structures & Algorithms/string-encode-and-decode/submission-1.cpp
class Solution {
public:

    string encode(vector<string>& strs) {
        string res = "";
        for (auto i : strs) {
            string length = to_string(i.size());
            res += length + "_" + i;
        }
        cout << res;
        return res;
    }

    vector<string> decode(string s) {
        int i = 0;
        vector<string> res;
        while (i < s.size()) {
            string temp = "";
            string w = "";
            while (temp != "_") {
                w += s[i];
                i++;
                temp = s[i];
            }
            int wordSize = stoi(w);
            string word = "";
            i++;
            for (int j = 0; j < wordSize; j++) {
                word += s[i];
                i++;
            }
            res.push_back(word);
            word = "";
        }
        return res;
    }
};

File: Data Structures & Algorithms/decode-ways/submission-1.py
class Solution:
    def numDecodings(self, s: str) -> int:
        n = len(s)
        res = {n: 1}
        def dp(idx):
            if idx in res:
                return res[idx]
            if s[idx] == "0":
                return 0
            result = dp(idx+1)
            if idx + 1 < n and (s[idx] == "1" or (s[idx] == "2" and s[idx+1] in "0123456")):
                result += dp(idx+2)
                res[idx] = result
            return result
        
        return dp(0)